File: crates/prova-core/testdata/describe.lua
-- describe: label-only subgrouping. Bare prova.test inside a top-level describe nests under the
-- label; describes nest; the GroupBuilder form takes a builder. No new fixture scope.

prova.describe("math", function()
  prova.test("adds", function(t)
    t:expect(1 + 1):equals(2)
  end)

  prova.describe("nested", function()
    prova.test("multiplies", function(t)
      t:expect(2 * 3):equals(6)
    end)
    -- test_each also nests under the ambient label.
    prova.test_each("squares {n}", { { n = 2, want = 4 } }, function(t, case)
      t:expect(case.n * case.n):equals(case.want)
    end)
  end)
end)

-- After the describe body returns, the ambient parent pops back to the file root.
prova.test("top level again", function(t)
  t:expect(true):is_truthy()
end)

-- GroupBuilder:describe — a labeling subgroup inside a group, using the builder.
prova.group("outer", function(g)
  g:describe("subsection", function(sub)
    sub:test("runs", function(t)
      t:expect("ok"):equals("ok")
    end)
  end)
end)

File: crates/prova-core/tests/describe.rs
use std::path::PathBuf;

use prova_core::{discover_path, run_path, NullReporter};

fn testdata(file: &str) -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("testdata")
        .join(file)
}

/// Every test under (possibly nested) `describe` labels runs and passes: 2 in `math`/`math › nested`
/// + 1 test_each + 1 top-level + 1 in the group's `describe` = 5.
#[test]
fn describe_tests_all_run() {
    let mut reporter = NullReporter;
    let summary = run_path(&testdata("describe.lua"), &mut reporter).expect("run describe.lua");
    assert_eq!(summary.passed, 5, "all describe-nested tests run");
    assert_eq!(summary.failed, 0, "failed");
}

/// `describe` labels appear in the reported path, nest, and pop back to the root afterward.
#[test]
fn describe_nests_labels_in_paths() {
    let names = discover_path(&testdata("describe.lua")).expect("discover describe.lua");
    assert!(
        names.iter().any(|n| n.contains("math › adds")),
        "expected 'math › adds', got: {names:?}"
    );
    assert!(
        names.iter().any(|n| n.contains("math › nested › multiplies")),
        "expected nested label path, got: {names:?}"
    );
    assert!(
        names.iter().any(|n| n.contains("math › nested › squares 2")),
        "expected test_each nested under describe, got: {names:?}"
    );
    // The post-describe test is back at the root (only the file-stem ancestor, no 'math').
    let top = names
        .iter()
        .find(|n| n.ends_with("top level again"))
        .expect("top-level test present");
    assert!(
        !top.contains("math"),
        "the post-describe test must pop back to root, got: {top:?}"
    );
    assert!(
        names.iter().any(|n| n.contains("outer › subsection › runs")),
        "expected GroupBuilder:describe nesting, got: {names:?}"
    );
}
